# Getting Started with R

[Original source](https://cengel.github.io/R-intro/)

## Creating objects in R

doesn't print anything

In [6]:
weight_kg <- 55

but putting parenthesis around the call prints the value of `weight_kg`

In [4]:
(weight_kg <- 55)

[1] 55

and so does typing the name of the object

In [5]:
weight_kg 

[1] 55

For instance, we may want to convert this weight into pounds (weight in pounds is 2.2 times the weight in kg):

In [7]:
2.2 * weight_kg

[1] 121

We can also change a variable’s value by assigning it a new one:

In [8]:
weight_kg <- 57.5
2.2 * weight_kg

[1] 126.5

let’s store the weight in pounds in a new variable,

In [9]:
weight_lb <- 2.2 * weight_kg

Challenge
What do you think is the current content of the object weight_lb? 126.5 or 220?

In [10]:
weight_lb

[1] 126.5

### Functions and their arguments

A function usually gets one or more inputs called arguments. Functions often (but not always) return a value.

An example of a function call is:

In [15]:
a <- 3.1416
b <- round(a)
b

[1] 3

We can use `args(function)` or look at the help for this function using `?function`.

In [13]:
args(round)

function (x, digits = 0) 
NULL

In [14]:
?round

Round                   package:base                   R Documentation

_R_o_u_n_d_i_n_g _o_f _N_u_m_b_e_r_s

_D_e_s_c_r_i_p_t_i_o_n:

     ‘ceiling’ takes a single numeric argument ‘x’ and returns a
     numeric vector containing the smallest integers not less than the
     corresponding elements of ‘x’.

     ‘floor’ takes a single numeric argument ‘x’ and returns a numeric
     vector containing the largest integers not greater than the
     corresponding elements of ‘x’.

     ‘trunc’ takes a single numeric argument ‘x’ and returns a numeric
     vector containing the integers formed by truncating the values in
     ‘x’ toward ‘0’.

     ‘round’ rounds the values in its first argument to the specified
     number of decimal places (default 0).

     ‘signif’ rounds the values in its first argument to the specified
     number of significant digits.

_U_s_a_g_e:

     ceiling(x)
     floor(x)
     trunc(x, ...)
     
     round(x, digits = 0)
     signif(x, digits = 6)
     
_A_r_g_u_m_e_n_t_s:

       x: a numeric vector.  Or, for ‘round’ and ‘signif’, a complex
          vector.

  digits: integer indicating the number of decimal places (‘round’) or
          significant digits (‘signif’) to be used.  Negative values
          are allowed (see ‘Details’).

     ...: arguments to be passed to methods.

_D_e_t_a_i_l_s:

     These are generic functions: methods can be defined for them
     individually or via the ‘Math’ group generic.

     Note that for rounding off a 5, the IEC 60559 standard is expected
     to be used, ‘_go to the even digit_’.  Therefore ‘round(0.5)’ is
     ‘0’ and ‘round(-1.5)’ is ‘-2’.  However, this is dependent on OS
     services and on representation error (since e.g. ‘0.15’ is not
     represented exactly, the rounding rule applies to the represented
     number and not to the printed number, and so ‘round(0.15, 1)’
     could be either ‘0.1’ or ‘0.2’).

     Rounding to a negative number of digits means rounding to a power
     of ten, so for example ‘round(x, digits = -2)’ rounds to the
     nearest hundred.

     For ‘signif’ the recognized values of ‘digits’ are ‘1...22’, and
     non-missing values are rounded to the nearest integer in that
     range.  Complex numbers are rounded to retain the specified number
     of digits in the larger of the components.  Each element of the
     vector is rounded individually, unlike printing.

     These are all primitive functions.

_S_4 _m_e_t_h_o_d_s:

     These are all (internally) S4 generic.

     ‘ceiling’, ‘floor’ and ‘trunc’ are members of the ‘Math’ group
     generic.  As an S4 generic, ‘trunc’ has only one argument.

     ‘round’ and ‘signif’ are members of the ‘Math2’ group generic.

_W_a_r_n_i_n_g:

     The realities of computer arithmetic can cause unexpected results,
     especially with ‘floor’ and ‘ceiling’.  For example, we ‘know’
     that ‘floor(log(x, base = 8))’ for ‘x = 8’ is ‘1’, but ‘0’ has
     been seen on an R platform.  It is normally necessary to use a
     tolerance.

_R_e_f_e_r_e_n_c_e_s:

     Becker, R. A., Chambers, J. M. and Wilks, A. R. (1988) _The New S
     Language_.  Wadsworth & Brooks/Cole.

_S_e_e _A_l_s_o:

     ‘as.integer’.

_E_x_a_m_p_l_e_s:

     round(.5 + -2:4) # IEEE rounding: -2  0  0  2  2  4  4
     ( x1 <- seq(-2, 4, by = .5) )
     round(x1) #-- IEEE rounding !
     x1[trunc(x1) != floor(x1)]
     x1[round(x1) != floor(x1 + .5)]
     (non.int <- ceiling(x1) != floor(x1))
     
     x2 <- pi * 100^(-1:3)
     round(x2, 3)
     signif(x2, 3)

In [111]:
??round

R Information

Vignettes with name or keyword or title matching ‘round’ using fuzzy
matching:


survival::tiedtimes     Roundoff error and tied times


Type 'vignette("FOO", package="PKG")' to inspect entries 'PKG::FOO'.



Demos with name or title matching ‘round’ using fuzzy matching:


sp::webmap              OpenStreetMap or google map backgrounds in
                        sp::plot and sp::spplot, using ggmap or
                        RgoogleMaps maps


Type 'demo(PKG::FOO)' to run demonstration 'PKG::FOO'.



Help files with alias or concept or title matching ‘round’ using fuzzy
matching:


FactoMineR::plotellipses
                        Draw confidence ellipses around the categories
Matrix::dgCMatrix-class
                        Compressed, sparse, column-oriented numeric
                        matrices
  Aliases: round,dgCMatrix,numeric-method
TTR::PBands             Construct (optionally further smoothed and
                        centered ) volatility bands around prices

We see that if we want a different number of digits, we can type digits=2 or however many we want.

In [16]:
round(3.14159, digits = 2)

[1] 3.14

If you provide the arguments in the exact same order as they are defined you don’t have to name them:

In [17]:
round(3.14159, 2)

[1] 3.14

And if you do name the arguments, you can switch their order:

In [19]:
round(digits = 2, x = 3.14159)

[1] 3.14

* **NOTE**: R evaluates function arguments in three steps: **first**, _by exact matching_ on argument name, **then** _by partial matching_ on argument name, and **finally** _by position_.

## Vectors and data types

A vector is the most common and basic data type in R, and is pretty much the workhorse of R. A vector is composed by a series of values, which can be either numbers or characters.
Exists 6 **atomic vector types** (or data types) that R uses:
* `"character"`
* `"numeric"`
* `"logical"`
* `"integer"` for integer numbers (e.g., 2L, the L indicates to R that it’s an integer)
* `"complex"` to represent complex numbers with real and imaginary parts (e.g., 1 + 4i) 
* `"raw"`

We can assign a series of values to a vector using the `c()` function.

In [20]:
weight_g <- c(21, 34, 39, 54, 55)
weight_g

[1] 21 34 39 54 55

`length()` tells you how many elements are in a particular vector:

In [21]:
length(weight_g)

[1] 5

`class()` indicates the class (the type of element) of an object:

In [23]:
class(weight_g)

[1] "numeric"

`str()` provides an overview of the structure of an object and its elements:

In [24]:
str(weight_g)

 num [1:5] 21 34 39 54 55


You can use the `c()` function to add other elements to your vector:

In [25]:
weight_g <- c(weight_g, 90) # add to the end of the vecto
weight_g

[1] 21 34 39 54 55 90

In [26]:
weight_g <- c(30, weight_g) # add to the beginning of the vector
weight_g

[1] 30 21 34 39 54 55 90

A vector can also contain characters:

In [27]:
animals <- c("mouse", "rat", "dog", "bear")
class(animals)

[1] "character"

we will introduce a vector with logical values (the boolean data type).

In [28]:
has_tail <- c(TRUE, TRUE, TRUE, FALSE)
has_tail 

[1]  TRUE  TRUE  TRUE FALSE

We’ve seen that atomic vectors can be of type character, numeric, integer, and logical. But what happens if we try to mix these types in a single vector?

Objects of different types get converted into a single, shared type within a vector. In R, we call converting objects from one class into another class **coercion**. These conversions happen according to a hierarchy, whereby some types get preferentially coerced into other types.

In [33]:
num_char <- c(1, 2, 3, 'a')
class(num_char)

[1] "character"

In [34]:
num_logical <- c(1, 2, 3, TRUE)
class(num_logical)

[1] "numeric"

In [35]:
char_logical <- c('a', 'b', 'c', TRUE)
class(char_logical)

[1] "character"

In [36]:
tricky <- c(1, 2, 3, '4')
class(tricky)

[1] "character"

## Subsetting vectors

In [38]:
animals <- c("mouse", "rat", "dog", "bear")
class(animals)

[1] "character"

In [43]:
animals[0]

character(0)

In [42]:
animals[1]

[1] "mouse"

In [40]:
animals[c(3, 2)]

[1] "dog" "rat"

`':'` is a special function that creates numeric vectors of integers in increasing or decreasing order

In [44]:
animals[2:4]

[1] "rat"  "dog"  "bear"

In [45]:
animals[4:2]

[1] "bear" "dog"  "rat"

You can exclude elements of a vector using the `“-”` sign:

In [46]:
animals[-2]

[1] "mouse" "dog"   "bear"

We can also repeat the indices to create an object with more elements than the original one:

In [47]:
more_animals <- animals[c(1, 2, 3, 2, 1, 1, 1, 1, 1, 4)]
more_animals

[1] "mouse" "rat"   "dog"   "rat"   "mouse" "mouse" "mouse" "mouse" "mouse"
[10] "bear"

* **NOTE**: R indices start at 1. Programming languages like Fortran, MATLAB, Julia, and R start counting at 1, because that’s what human beings typically do. Languages in the C family (including C++, Java, Perl, and Python) count from 0 because that’s simpler for computers to do.

### Conditional subsetting

Another common way of subsetting is by using a logical vector. `TRUE` will select the element with the same index, while `FALSE` will not.

In [38]:
animals <- c("mouse", "rat", "dog", "bear")
class(animals)

[1] "character"

In [48]:
has_tail <- c(TRUE, TRUE, TRUE, FALSE)
has_tail 

[1]  TRUE  TRUE  TRUE FALSE

In [49]:
animals[has_tail]

[1] "mouse" "rat"   "dog"

In [47]:
more_animals <- animals[c(1, 2, 3, 2, 1, 1, 1, 1, 1, 4)]
more_animals

[1] "mouse" "rat"   "dog"   "rat"   "mouse" "mouse" "mouse" "mouse" "mouse"
[10] "bear"

In [50]:
more_animals[has_tail]

[1] "mouse" "rat"   "dog"   "mouse" "mouse" "mouse" "mouse" "bear"

To search for certain strings in a vector. One could use the `“or”` operator `|` to test for equality to multiple values, but this can quickly become tedious. The function `%in%` allows you to test if any of the elements of a search vector are found:

In [51]:
animals[animals == "bear" | animals == "rat"] # returns both rat and cat

[1] "rat"  "bear"

In [52]:
animals %in% c("rat", "cat", "dog", "duck", "goat")

[1] FALSE  TRUE  TRUE FALSE

In [54]:
weight_g <- c(21, 34, 39, 54, 55)
weight_g

[1] 21 34 39 54 55

In [55]:
weight_g > 50    # will return logicals with TRUE for the indices that meet the condition

[1] FALSE FALSE FALSE  TRUE  TRUE

You can combine multiple tests using `&` (both conditions are true,`AND`) or `|` (at least one of the conditions is true, `OR`):

In [58]:
weight_g[weight_g < 30 | weight_g > 50]

[1] 21 54 55

In [59]:
weight_g[weight_g >= 30 & weight_g == 21]

numeric(0)

## Missing data

As `R` was designed to analyze datasets, it includes the concept of missing data (which is uncommon in other programming languages). Missing data are represented in vectors as `NA`.

When doing operations on numbers, most functions will return `NA` if the data you are working with include missing values. This feature makes it harder to overlook the cases where you are dealing with missing data.

In [61]:
heights <- c(2, 4, 4, NA, 6)
heights

[1]  2  4  4 NA  6

In [62]:
max(heights)

[1] NA

In [63]:
sum(heights)

[1] NA

You can add the argument `na.rm=TRUE` to calculate the result while ignoring the missing values.

In [64]:
max(heights, na.rm = TRUE)

[1] 6

In [65]:
sum(heights, na.rm = TRUE)

[1] 16

### Dealing with NA

If your data include missing values, you may want to become familiar with the functions:
* `is.na()`, 
* `na.omit()`, and 
* `complete.cases()`

In [69]:
# Extract elements which are not missing values.
heights[!is.na(heights)] 

[1] 2 4 4 6

In [70]:
# Returns the object with incomplete cases removed. The returned object is atomic.
na.omit(heights)

[1] 2 4 4 6
attr(,"na.action")
[1] 4
attr(,"class")
[1] "omit"

In [75]:
# Extract elements which are complete cases.
heights[complete.cases(heights)]

[1] 2 4 4 6

Challenge
Using this vector of length measurements, create a new vector with the NAs removed.

In [77]:
lengths <- c(10,24,NA,18,NA,20)
lengths

[1] 10 24 NA 18 NA 20

In [79]:
na.omit(lengths)

[1] 10 24 18 20
attr(,"na.action")
[1] 3 5
attr(,"class")
[1] "omit"

In [80]:
median(lengths, na.rm = TRUE)

[1] 19

In [81]:
mean(lengths, na.rm = TRUE)

[1] 18

## Common R Data Structures

Vectors are one of the many data structures that R uses. Other important ones are matrices (`matrix`), tables (`data.frame`), lists (`list`), and factors (`factor`).

### Matrix

To construct a matrix, we use a function conveniently called `matrix()`.

In [82]:
y <- matrix(1:20, nrow=5,ncol=4) # generates 5 x 4 numeric matrix
y

[,1] [,2] [,3] [,4]
[1,] 1     6   11   16  
[2,] 2     7   12   17  
[3,] 3     8   13   18  
[4,] 4     9   14   19  
[5,] 5    10   15   20

Subset a matrix with `[row , column]`:

In [85]:
y[,4]       # 4th column of matrix

[1] 16 17 18 19 20

In [86]:
y[3,]       # 3rd row of matrix

[1]  3  8 13 18

In [87]:
y[2:4,1:3]  # rows 2,3,4 of columns 1,2,3

[,1] [,2] [,3]
[1,] 2    7    12  
[2,] 3    8    13  
[3,] 4    9    14

### List

Lists can have elements of any type. Here is how we construct lists. You may have guessed that to construct a list, we use the `list()` function:

In [89]:
myl <- list(id="ID_1", a_vector=animals, a_matrix=y, age=5.3) # example of a list with 4 components
myl

$id
[1] "ID_1"

$a_vector
[1] "mouse" "rat"   "dog"   "bear" 

$a_matrix
     [,1] [,2] [,3] [,4]
[1,]    1    6   11   16
[2,]    2    7   12   17
[3,]    3    8   13   18
[4,]    4    9   14   19
[5,]    5   10   15   20

$age
[1] 5.3

In [93]:
myl[[2]] # 2nd component of the list

[1] "mouse" "rat"   "dog"   "bear"

In [96]:
myl[["id"]] # component named id in list

[1] "ID_1"

### Data frame

Data frames in R are a special case of lists, as they can have elements of any type, but they have to all be of the same length.

A data frame is the most common way of storing tabular data in R and something you will likely deal with a lot. As a first approximation, which holds true, probably in the most cases, you can really think of it as a table or a spreadsheet.

In [97]:
mydf <- data.frame(ID=c(1:4),
                   Color=c("red", "white", "red", NA),
                   Passed=c(TRUE,TRUE,TRUE,FALSE),
                   Weight=c(99, 54, 85, 70),
                   Height=c(1.78, 1.67, 1.82, 1.59))

mydf

ID Color Passed Weight Height
1 1  red    TRUE  99     1.78  
2 2  white  TRUE  54     1.67  
3 3  red    TRUE  85     1.82  
4 4  NA    FALSE  70     1.59

Challenge
1. Create a data frame that holds the following information for yourself, your right and your left neighbor:
    * first name
    * last name
    * lucky number
2. There are a few mistakes in this hand-crafted data.frame, can you spot and fix them? Don’t hesitate to experiment!

In [103]:
df <- data.frame(ID=c(1:2),
                 first_name=c('Jacobo','Gerardo'), 
                 last_name=c('León', 'León'), 
                 lucky_number=c(1, 3))
df

ID first_name last_name lucky_number
1 1  Jacobo     León      1           
2 2  Gerardo    León      3

In [108]:
animal_data <- data.frame(animal=c("dog", "cat", "sea cucumber", "sea urchin"),
                          feel=c("furry", "squishy", "spiny", NA),
                          weight=c(45, 8, 1.1, 0.8))

animal_data

animal       feel    weight
1 dog          furry   45.0  
2 cat          squishy  8.0  
3 sea cucumber spiny    1.1  
4 sea urchin   NA       0.8

## Extending R base functionality

Install from the R console like this

In [109]:
install.packages("lubridate", dependencies = TRUE)

Warning message:
“unable to access index for repository https://cran.r-project.org/src/contrib:
  cannot open URL 'https://cran.r-project.org/src/contrib/PACKAGES'”Warning message:
“package ‘lubridate’ is not available (for R version 3.4.3)”

### Make use of the installed packages

In order to actually use commands from the installed packages you also will need to load the installed packages:
* `library(lubridate)`
* `require(lubridate)`

The difference between the two is that `library` will result in an error, if the library does not exist, whereas `require` will result in a warning.

### How to ask for help

To share an object with someone else, if it’s relatively small, you can use the function `dput()`. It will output R code that can be used to recreate the exact same object as the one in memory:

In [112]:
dput(head(iris)) # iris is an example data frame that comes with R and head() is a function that returns the first part of the data frame

structure(list(Sepal.Length = c(5.1, 4.9, 4.7, 4.6, 5, 5.4), 
    Sepal.Width = c(3.5, 3, 3.2, 3.1, 3.6, 3.9), Petal.Length = c(1.4, 
    1.4, 1.3, 1.5, 1.4, 1.7), Petal.Width = c(0.2, 0.2, 0.2, 
    0.2, 0.2, 0.4), Species = structure(c(1L, 1L, 1L, 1L, 1L, 
    1L), .Label = c("setosa", "versicolor", "virginica"), class = "factor")), .Names = c("Sepal.Length", 
"Sepal.Width", "Petal.Length", "Petal.Width", "Species"), row.names = c(NA, 
6L), class = "data.frame")


Last, but certainly not least, always include the output of sessionInfo() as it provides critical information about your platform, the versions of R and the packages that you are using, and other information that can be very helpful to understand your problem.

In [115]:
sessionInfo()

R version 3.4.3 (2017-11-30)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X El Capitan 10.11.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib

locale:
[1] C/UTF-8/C/C/C/C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] compiler_3.4.3  magrittr_1.5    IRdisplay_0.5.0 pbdZMQ_0.3-3   
 [5] tools_3.4.3     htmltools_0.3.6 base64enc_0.1-3 crayon_1.3.4   
 [9] Rcpp_0.12.17    uuid_0.1-2      stringi_1.2.3   IRkernel_0.8.11
[13] jsonlite_1.5    stringr_1.3.1   digest_0.6.15   repr_0.15.0    
[17] evaluate_0.10.1